packages/embedder/mergeAsyncIterators.ts

```diff
@@ -0,0 +1,101 @@
+// can remove PromiseCapability after TS v5.4.2
+type PromiseCapability<T> = {
+  resolve: (value: T) => void
+  reject: (reason?: any) => void
+  promise: Promise<T>
+}
+
+// typescript intrinsics for iterables don't work like the do for promises yet, please ignore hackery
+type UnYield<T> = T extends IteratorYieldResult<infer U> ? U : never
+type UnIt<T extends AsyncIterator<any>> = UnYield<Awaited<ReturnType<T['next']>>>
+
+// Promise.race has a memory leak
+// To avoid: https://github.com/tc39/proposal-async-iterator-helpers/issues/15#issuecomment-1937011820
+export function mergeAsyncIterators<
```

Comment:
+1 

I encourage you to completely ignore this comment but file it away for some future date.


I think this pattern might be useful enough that it warrants becoming its own little library. Rather than work on just two iterables, I think you should extended its logic to operate on n-iterables.

The dance you do between awaiting on the promise (to allow the iterators to run) and adding to the queue is very clever.

The old man in me thinks of:

   * In UNIX, we had select in order to tell us when one or more non-blocking file or socket descriptors was ready for reading, writing (or had errored)
   * In Windows NT, we had `WaitForMultipleObjects` that would await for multiple kinds of handles had an event (such as file handles, event handles, mutexes, and semaphores)
   * Go has a `select { }` statement that does almost exactly what you've created here. Given multiple go channels, it awaits on each of them and runs a code block for the one that has data

Reply:
renamed the functions to make reading it a little clearer.
supported n async iterables. that way we can run n job queue workers & still get good types.